pkg/controller/deployment/util/deployment_util.go

```diff
@@ -415,14 +415,25 @@ func SetReplicasAnnotations(rs *extensions.ReplicaSet, desiredReplicas, maxRepli
 
 // MaxUnavailable returns the maximum unavailable pods a rolling deployment can take.
 func MaxUnavailable(deployment extensions.Deployment) int32 {
-	if !IsRollingUpdate(&deployment) {
+	if !IsRollingUpdate(&deployment) || *(deployment.Spec.Replicas) == 0 {
```

Comment:
IMHO the logic should exist in `func MinAvailable` rather than `func MaxUnavailable`. We have `MaxUnavailable` field in the spec, and `func MaxUnavailable` should return its callers the `int32` value of that field, regardless of the number of replicas; otherwise the returned value is surprising to the callers. OTOH, `func MinAvailable` should return a reasonable min number of replicas that should be available, taking the number of replicas into consideration.